core/pubnub_coreapi_ex.c

```diff
@@ -241,3 +242,67 @@ enum pubnub_res pubnub_history_ex(pubnub_t*                     pb,
     pubnub_mutex_unlock(pb->monitor);
     return rslt;
 }
+
+struct pubnub_set_state_options pubnub_set_state_defopts(void)
+{
+    struct pubnub_set_state_options options;
+
+    options.user_id = NULL;
+    options.channel_group = NULL;
+    options.heartbeat = false;
+
+    return options;
+}
+
+static enum pubnub_res heartbeat_with_state(pubnub_t *pb,
+        const char *channel,
+        const char *state,
+        struct pubnub_set_state_options opts)
+{
+    enum pubnub_res rslt;
+
+    PUBNUB_ASSERT(pb_valid_ctx_ptr(pb));
+
+    pubnub_mutex_lock(pb->monitor);
+    if (!pbnc_can_start_transaction(pb)) {
+        pubnub_mutex_unlock(pb->monitor);
+        return PNR_IN_PROGRESS;
+    }
+
+    rslt = pbauto_heartbeat_prepare_channels_and_ch_groups(pb, &channel, &opts.channel_group);
+    if (rslt != PNR_OK) {
+        return rslt;
+    }
+    
+    size_t state_lenght = strlen(state);
+    if (pb->core.state == NULL) {
+        pb->core.state = (char*)malloc(state_lenght + 1);
+    } else if (strlen(pb->core.state) < state_lenght) {
+        pb->core.state = (char*)realloc((char*)pb->core.state, state_lenght + 1);
+    }
+    strcpy((char*)pb->core.state, state);
```

Comment:
heartbeat=true, is not giving expected state in the URL.
with existing set_state, heartbeat URL is `/v2/presence/sub-key/demo/channel/hello_world/heartbeat?pnsdk=Windows-PubNub-C-core%2F4.0.6&uuid=6e581826-3787-4d70-a465-d749226345f5&state=%7B%22hello_world%22%3A%7B%22x%22%3A5%7D%7D&auth=danaske`
Note the format of state {<channel>:<state provided by user>}

with heartbeat=true, the URL I see is `/v2/presence/sub-key/demo/channel/hello_world/heartbeat?pnsdk=Windows-PubNub-C-core%2F4.0.6&uuid=91fa116f-871b-459a-9009-b3997eb6b041&state=%7B%22x%22%3A5%7D&auth=danaske`
which is missing channel name in the state.

Reply:
Changed as we discuss at DM